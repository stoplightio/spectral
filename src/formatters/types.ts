import { IRuleResult } from '../types';

export type Formatter = (results: IRuleResult[]) => string;
