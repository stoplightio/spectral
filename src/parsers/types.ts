import { GetLocationForJsonPath, IParserResult } from '@stoplight/types';

export interface IParser<R extends IParserResult = IParserResult<unknown>> {
  parse(input: string): R;
  getLocationForJsonPath: GetLocationForJsonPath<R>;
  trapAccess<T extends object = object>(obj: T): T;
}
