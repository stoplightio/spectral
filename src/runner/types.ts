import type { DocumentInventory } from '../documentInventory';
import type { Ruleset } from '../ruleset/ruleset';
import { ISpectralDiagnostic } from '../types';

export interface IRunnerInternalContext {
  ruleset: Ruleset;
  documentInventory: DocumentInventory;
  results: ISpectralDiagnostic[];
  promises: Array<Promise<void>>;
}
