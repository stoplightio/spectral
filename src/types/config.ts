import { HumanReadableDiagnosticSeverity } from './rule';

export type FailSeverity = HumanReadableDiagnosticSeverity;

export enum OutputFormat {
  JSON = 'json',
  STYLISH = 'stylish',
  JUNIT = 'junit',
  HTML = 'html',
  TEXT = 'text',
  TEAMCITY = 'teamcity',
}

export interface ILintConfig {
  encoding: string;
  format: OutputFormat;
  output?: string;
  resolver?: string;
  ruleset?: string[];
  skipRule?: string[];
  ignoreUnknownFormat: boolean;
  showUnmatchedGlobs: boolean;
  failOnUnmatchedGlobs: boolean;
  verbose?: boolean;
  quiet?: boolean;
}
