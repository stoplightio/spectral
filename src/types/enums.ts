export enum RuleType {
  VALIDATION = 'validation',
  STYLE = 'style',
}

export enum RuleFunction {
  ALPHABETICAL = 'alphabetical',
  CASING = 'casing',
  LENGTH = 'length',
  PATTERN = 'pattern',
  ENUM = 'enumeration',
  SCHEMA = 'schema',
  SCHEMAPATH = 'schemaPath',
  TRUTHY = 'truthy',
  XOR = 'xor',
  TYPED_ENUM = 'typedEnum',
}
