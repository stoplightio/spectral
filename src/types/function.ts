import { JsonPath } from '@stoplight/types';
import { DocumentInventory } from '../documentInventory';

export type IFunction<O = any> = (
  targetValue: any,
  options: O,
  paths: IFunctionPaths,
  otherValues: IFunctionValues,
) => void | IFunctionResult[];

export interface IFunctionPaths {
  given: JsonPath;
  target?: JsonPath;
}

export interface IFunctionValues {
  original: any;
  given: any;
  documentInventory: DocumentInventory;
}

export interface IFunctionResult {
  message: string;
  path?: JsonPath;
}
