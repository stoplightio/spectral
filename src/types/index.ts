export * from './spectral';
export * from './function';
