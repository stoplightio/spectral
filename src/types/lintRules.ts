import { IRuleDefinitionBase } from './index';

export interface IRuleParam {
  properties: string | string[];
}

export interface IRuleStringParam extends IRuleParam {
  value: string;
}

export interface IRuleNumberParam {
  value: number;
  property?: string;
}

export interface IAlphaRuleParam extends IRuleParam {
  // if sorting objects, use key for comparison
  keyedBy?: string;
}

export interface IRulePatternParam {
  // object key to apply rule to
  property: string;

  // value to use for rule
  value: string;

  // value to omit from regex matching
  omit?: string;

  // value to split the property on prior to performing regex matching
  split?: string;
}

export interface ITruthyRule extends IRuleDefinitionBase {
  type: 'truthy';

  // key(s) of object that should evaluate as 'truthy' (considered true in a
  // boolean context)
  truthy: string | string[];

  properties?: number;
}

export interface IOrRule extends IRuleDefinitionBase {
  type: 'or';

  // test to verify if any of the provided keys are present in object
  or: string[];
}

export interface IXorRule extends IRuleDefinitionBase {
  type: 'xor';

  // test to verify if one (but not all) of the provided keys are present in
  // object
  xor: string[];
}

export interface IMaxLengthRule extends IRuleDefinitionBase {
  type: 'maxLength';

  // verify property is under a specified number of characters
  maxLength: IRuleNumberParam;
}

export interface IAlphaRule extends IRuleDefinitionBase {
  type: 'alphabetical';

  // verify property is within alphabetical order
  alphabetical: IAlphaRuleParam;
}

export interface INotEndWithRule extends IRuleDefinitionBase {
  type: 'notEndWith';

  // verify property does not end with string
  notEndWith: IRulePatternParam;
}

export interface INotContainRule extends IRuleDefinitionBase {
  type: 'notContain';

  // verify property does not contain value
  notContain: IRuleStringParam;
}

export interface IPatternRule extends IRuleDefinitionBase {
  type: 'pattern';

  // run regex match
  pattern: IRulePatternParam;
}

export interface IFunctionRule extends IRuleDefinitionBase {
  type: 'function';
  function: string; // eventually allow js functions here? or references to one
  args?: any; //need to think about this a little
}

export type LintRule =
  | ITruthyRule
  | IOrRule
  | IXorRule
  | IMaxLengthRule
  | IAlphaRule
  | INotEndWithRule
  | INotContainRule
  | IPatternRule
  | IFunctionRule;
