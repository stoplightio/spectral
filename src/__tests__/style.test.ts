import { Spectral } from '../index';
import { IRuleResult, IRuleset, Rule, RuleFunction, RuleType } from '../types';

const applyRuleToObject = (r: Rule, o: object): IRuleResult[] => {
  const cfg: IRuleset[] = [
    {
      rules: {
        testing: {
          'test:rule': r,
        },
      },
    },
  ];
  const s = new Spectral({ rulesets: cfg });
  return s.run({ target: o, spec: 'testing' });
};

describe('lint', () => {
  describe('rules', () => {
    describe('truthy', () => {
      test('returns result if value is not present', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.TRUTHY,
              path: '$.info',
              enabled: true,
              summary: '',
              input: {
                properties: 'something-not-present',
              },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0.0',
                title: 'Swagger Petstore',
                termsOfService: 'http://swagger.io/terms/',
              },
            }
          )
        ).toMatchSnapshot();
      });

      test('doesnt return result if value is present', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.TRUTHY,
              path: '$.info',
              enabled: true,
              summary: '',
              input: { properties: 'version' },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0.0',
                title: 'Swagger Petstore',
                termsOfService: 'http://swagger.io/terms/',
              },
            }
          )
        ).toMatchSnapshot();
      });
    });

    describe('alphabetical', () => {
      test('returns result if values are not alphabetized', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.ALPHABETICAL,
              path: '$.info',
              enabled: true,
              summary: '',
              input: {
                properties: 'tags',
                keyedBy: 'name',
              },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0.0',
                title: 'Swagger Petstore',
                tags: [{ name: 'Far', description: 'bar' }, { name: 'Boo', description: 'foo' }],
              },
            }
          )
        ).toMatchSnapshot();
      });

      test('dont return result if values are alphabetized', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.ALPHABETICAL,
              path: '$.info',
              enabled: true,
              summary: '',
              input: {
                properties: 'tags',
                keyedBy: 'name',
              },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0.0',
                title: 'Swagger Petstore',
                tags: [{ name: 'Boo', description: 'bar' }, { name: 'Far', description: 'foo' }],
              },
            }
          )
        ).toMatchSnapshot();
      });
    });

    describe('or', () => {
      test('returns result if no properties are present', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.OR,
              path: '$.info',
              enabled: true,
              summary: '',
              input: { properties: ['something-not-present', 'something-else-not-present'] },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0.0',
                title: 'Swagger Petstore',
              },
            }
          )
        ).toMatchSnapshot();
      });

      test('dont returns results if any properties are present', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.OR,
              path: '$.info',
              enabled: true,
              summary: '',
              input: { properties: ['version', 'something-else-not-present'] },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0.0',
                title: 'Swagger Petstore',
              },
            }
          )
        ).toMatchSnapshot();
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.OR,
              path: '$.info',
              enabled: true,
              summary: '',
              input: { properties: ['version', 'title', 'termsOfService'] },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0.0',
                title: 'Swagger Petstore',
                termsOfService: 'http://swagger.io/terms/',
              },
            }
          )
        ).toMatchSnapshot();
      });
    });

    describe('xor', () => {
      test('returns result if no properties are present', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.XOR,
              path: '$.info',
              enabled: true,
              summary: '',
              input: { properties: ['yada-yada', 'whatever'] },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0.0',
                title: 'Swagger Petstore',
                termsOfService: 'http://swagger.io/terms/',
              },
            }
          )
        ).toMatchSnapshot();
      });

      test('returns result if both properties are present', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.XOR,
              path: '$.info',
              enabled: true,
              summary: '',
              input: { properties: ['version', 'title'] },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0.0',
                title: 'Swagger Petstore',
                termsOfService: 'http://swagger.io/terms/',
              },
            }
          )
        ).toMatchSnapshot();
      });

      test('dont returns results if one of the properties are present', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.XOR,
              path: '$.info',
              enabled: true,
              summary: '',
              input: { properties: ['something', 'title'] },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0.0',
                title: 'Swagger Petstore',
                termsOfService: 'http://swagger.io/terms/',
              },
            }
          )
        ).toMatchSnapshot();
      });
    });

    describe('pattern', () => {
      test('returns result if pattern is not matched (on string)', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.PATTERN,
              path: '$.info',
              enabled: true,
              summary: '',
              input: {
                property: 'termsOfService',
                value: '^orange.*$',
              },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0.0',
                title: 'Swagger Petstore',
                termsOfService: 'http://swagger.io/terms/',
              },
            }
          )
        ).toMatchSnapshot();
      });

      test('returns result if pattern is not matched (on object keys)', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.PATTERN,
              path: '$.responses',
              enabled: true,
              summary: '',
              input: {
                property: '*',
                value: '^[0-9]+$',
              },
            },
            {
              responses: {
                '123': {
                  test: 'something',
                },
                '456avbas': {
                  test: 'something',
                },
                '789': {
                  test: 'something',
                },
              },
            }
          )
        ).toMatchSnapshot();
      });

      test('dont return result if pattern is matched (on string)', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.PATTERN,
              path: '$.info',
              enabled: true,
              summary: '',
              input: {
                property: 'termsOfService',
                value: '^http.*$',
              },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0.0',
                title: 'Swagger Petstore',
                termsOfService: 'http://swagger.io/terms/',
              },
            }
          )
        ).toMatchSnapshot();
      });

      test('dont return result if pattern is matched (on object keys)', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.PATTERN,
              path: '$.responses',
              enabled: true,
              summary: '',
              input: {
                property: '*',
                value: '^[0-9]+$',
              },
            },
            {
              responses: {
                '123': {
                  test: 'something',
                },
                '456': {
                  test: 'something',
                },
                '789': {
                  test: 'something',
                },
              },
            }
          )
        ).toMatchSnapshot();
      });
    });

    describe('notContain', () => {
      test('returns result if property contains value', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.NOT_CONTAIN,
              path: '$..*',
              enabled: true,
              summary: '',
              input: { properties: ['description'], value: '<script' },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0',
                title: 'To-do Demo',
                description:
                  "### Notes:\n\nThis OAS2 (Swagger 2) specification defines common models and responses, that other specifications may reference.\n\nFor example, check out the user poperty in the main.oas2 todo-partial model - it references the user model in this specification!\n\nLikewise, the main.oas2 operations reference the shared error responses in this common specification.\n\n<script>console.log('sup homie');</script>",
              },
            }
          )
        ).toMatchSnapshot();
      });

      test('dont return results if property doesnt contain value', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.NOT_CONTAIN,
              path: '$..*',
              enabled: true,
              summary: '',
              input: { properties: ['description'], value: '<script' },
            },
            {
              swagger: '2.0',
              info: {
                version: '1.0',
                title: 'To-do Demo',
                description:
                  '### Notes:\n\nThis OAS2 (Swagger 2) specification defines common models and responses, that other specifications may reference.\n\nFor example, check out the user poperty in the main.oas2 todo-partial model - it references the user model in this specification!\n\nLikewise, the main.oas2 operations reference the shared error responses in this common specification.',
              },
            }
          )
        ).toMatchSnapshot();
      });
    });

    describe('notEndWith', () => {
      test('return result if property ends with value', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.NOT_END_WITH,
              path: '$.servers',
              enabled: true,
              summary: '',
              input: { property: 'url', value: '/' },
            },
            {
              swagger: '2.0',
              servers: [
                {
                  url: 'http://localhost:5000/',
                  description: 'Development server',
                },
                {
                  url: 'https://rooms-staging.wework.com',
                  description: 'Staging server',
                },
              ],
            }
          )
        ).toMatchSnapshot();
      });

      test('dont return result if property doesnt end with value', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.NOT_END_WITH,
              path: '$.servers',
              enabled: true,
              summary: '',
              input: { property: 'url', value: '/' },
            },
            {
              swagger: '2.0',
              servers: [
                {
                  url: 'http://localhost:5000',
                  description: 'Development server',
                },
                {
                  url: 'https://rooms-staging.wework.com',
                  description: 'Staging server',
                },
              ],
            }
          )
        ).toMatchSnapshot();
      });
    });

    describe('maxLength', () => {
      test('return result if property is longer than value', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.MAX_LENGTH,
              path: '$..summary',
              enabled: true,
              description: 'summary should be short (description can be long)',
              summary: '',
              input: {
                value: 20,
              },
            },
            {
              paths: {
                '/rooms/{room_id}/reserve/': {
                  post: {
                    summary: 'Book Room Really fsdasddssdfgfdhdsafhsad fsad flong fjkdhfsds',
                  },
                },
              },
            }
          )
        ).toMatchSnapshot();
      });

      test('dont return result if property is shorter than value', () => {
        expect(
          applyRuleToObject(
            {
              type: RuleType.STYLE,
              function: RuleFunction.MAX_LENGTH,
              path: '$..summary',
              enabled: true,
              description: 'summary should be short (description can be long)',
              summary: '',
              input: {
                value: 20,
              },
            },
            {
              paths: {
                '/rooms/{room_id}/reserve/': {
                  post: {
                    summary: 'Book',
                  },
                },
              },
            }
          )
        ).toMatchSnapshot();
      });
    });
  });
});
