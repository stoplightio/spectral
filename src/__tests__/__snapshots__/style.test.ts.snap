// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lint rules alphabetical dont return result if values are alphabetized 1`] = `Array []`;

exports[`lint rules alphabetical returns result if values are not alphabetized 1`] = `
Array [
  Object {
    "description": "",
    "message": "expected Array [
  Object { name: 'Far', description: 'bar' },
  Object { name: 'Boo', description: 'foo' }
] to equal Array [
  Object { name: 'Boo', description: 'foo' },
  Object { name: 'Far', description: 'bar' }
] (at '0' -> name, A has 'Far' and B has 'Boo')",
    "name": "test:rule",
    "path": Array [
      "$",
      "info",
    ],
    "severity": 40,
    "severityLabel": "warn",
    "type": "style",
  },
]
`;

exports[`lint rules maxLength dont return result if property is shorter than value 1`] = `Array []`;

exports[`lint rules maxLength return result if property is longer than value 1`] = `
Array [
  Object {
    "description": "",
    "message": "expected 61 to be below or equal 20",
    "name": "test:rule",
    "path": Array [
      "$",
      "paths",
      "/rooms/{room_id}/reserve/",
      "post",
      "summary",
    ],
    "severity": 40,
    "severityLabel": "warn",
    "type": "style",
  },
]
`;

exports[`lint rules notContain dont return results if property doesnt contain value 1`] = `Array []`;

exports[`lint rules notContain returns result if property contains value 1`] = `
Array [
  Object {
    "description": "",
    "message": "expected '### Notes:\\\\n\\\\nThis OAS2 (Swagger 2) specification defines common models and responses, that other specifications may reference.\\\\n\\\\nFor example, check out the user poperty in the main.oas2 todo-partial model - it references the user model in this specification!\\\\n\\\\nLikewise, the main.oas2 operations reference the shared error responses in this common specification.\\\\n\\\\n<script>console.log(\\\\'sup homie\\\\');</script>' not to match /<script/ (false negative fail)",
    "name": "test:rule",
    "path": Array [
      "$",
      "info",
    ],
    "severity": 40,
    "severityLabel": "warn",
    "type": "style",
  },
]
`;

exports[`lint rules notEndWith dont return result if property doesnt end with value 1`] = `Array []`;

exports[`lint rules notEndWith return result if property ends with value 1`] = `
Array [
  Object {
    "description": "",
    "message": "expected 'http://localhost:5000/' not to end with '/' (false negative fail)",
    "name": "test:rule",
    "path": Array [
      "$",
      "servers",
    ],
    "severity": 40,
    "severityLabel": "warn",
    "type": "style",
  },
]
`;

exports[`lint rules or dont returns results if any properties are present 1`] = `Array []`;

exports[`lint rules or dont returns results if any properties are present 2`] = `Array []`;

exports[`lint rules or returns result if no properties are present 1`] = `
Array [
  Object {
    "description": "",
    "message": "expected false to be true",
    "name": "test:rule",
    "path": Array [
      "$",
      "info",
    ],
    "severity": 40,
    "severityLabel": "warn",
    "type": "style",
  },
]
`;

exports[`lint rules pattern dont return result if pattern is matched (on object keys) 1`] = `Array []`;

exports[`lint rules pattern dont return result if pattern is matched (on string) 1`] = `Array []`;

exports[`lint rules pattern returns result if pattern is not matched (on object keys) 1`] = `
Array [
  Object {
    "description": "",
    "message": ", but received: 456avbas",
    "name": "test:rule",
    "path": Array [
      "$",
      "responses",
    ],
    "severity": 40,
    "severityLabel": "warn",
    "type": "style",
  },
]
`;

exports[`lint rules pattern returns result if pattern is not matched (on string) 1`] = `
Array [
  Object {
    "description": "",
    "message": ", but received: http://swagger.io/terms/",
    "name": "test:rule",
    "path": Array [
      "$",
      "info",
    ],
    "severity": 40,
    "severityLabel": "warn",
    "type": "style",
  },
]
`;

exports[`lint rules truthy doesnt return result if value is present 1`] = `Array []`;

exports[`lint rules truthy returns result if value is not present 1`] = `
Array [
  Object {
    "description": "",
    "message": "expected Object {
  version: '1.0.0',
  title: 'Swagger Petstore',
  termsOfService: 'http://swagger.io/terms/'
} to have property 'something-not-present'",
    "name": "test:rule",
    "path": Array [
      "$",
      "info",
    ],
    "severity": 40,
    "severityLabel": "warn",
    "type": "style",
  },
]
`;

exports[`lint rules xor dont returns results if one of the properties are present 1`] = `Array []`;

exports[`lint rules xor returns result if both properties are present 1`] = `
Array [
  Object {
    "description": "",
    "message": "expected true undefined false",
    "name": "test:rule",
    "path": Array [
      "$",
      "info",
    ],
    "severity": 40,
    "severityLabel": "warn",
    "type": "style",
  },
]
`;

exports[`lint rules xor returns result if no properties are present 1`] = `
Array [
  Object {
    "description": "",
    "message": "expected false to be true",
    "name": "test:rule",
    "path": Array [
      "$",
      "info",
    ],
    "severity": 40,
    "severityLabel": "warn",
    "type": "style",
  },
]
`;
